### PR TITLE
DPL: Configuration integration

### PR DESCRIPTION
**Pull request by ktf**

### CODE REVIEW

**Comment by ktf** [comment]
@awegrzyn @teo @vascobarroso this should allow DPL integration with AliceO2Group/Configuration. Could you please have a look? Basically if you pass `--configuration uri` you can configure the parameters of your device using Configuration. There is a few rough edges, still:

* At the moment you cannot override options from the command line, if you are using e.g. a JSON file, only the JSON file values are used over the hardcoded defaults. Not sure if this is a bug or a feature, actually.
* There seem to be some quirks in configuration when dealing with ini files. See associated JIRA ticket.
* I've not tested consul.
* The GUI will show the default values when you use the `--configuration` option.
* Going back to Init() will not reload the options.

File: Framework/Core/src/ConfigurationOptionsRetriever.cxx [hunk 1]
```diff
@@ -34,7 +34,13 @@ ConfigurationOptionsRetriever::ConfigurationOptionsRetriever(std::vector<ConfigP
...
     mStore{}
 {
-  PropertyTreeHelpers::populate(schema, mStore, cfg->getRecursive(mainKey));
+  boost::property_tree::ptree in;
+  try {
+    in = cfg->getRecursive(mainKey);
+  } catch (...) {
+    in.clear();
+  }
+  PropertyTreeHelpers::populate(schema, mStore, in);
```
Comment (ktf) [inline Framework/Core/src/ConfigurationOptionsRetriever.cxx:43]: @awegrzyn Is there a way to detect if a given node exists?

**Comment by teo** [comment]
@ktf Does this also affect runtime device configuration? Can the same payload as `--configuration` be dropped into FMQ options and accessed equally by user code?
Ideally we would want to be able to start with no configuration at all, and provide everything at runtime.

**Comment by ktf** [comment]
This is for user code at the moment. So you basically want me to reconfigure the channels in the Init? Doesn't that have to be done in the FairMQ Control plugin?

**Comment by ktf** [comment]
Merging this as it should behave as usual when `--configuration` is not provided we can improve it further (e.g. by allowing multiple backends overriding each other) in as subsequent PR.

**Comment by teo** [comment]
@ktf what is "the Init" for you? Are you referring to the `INIT DEVICE`/`INIT TASK` SM path in the FairMQ SM or a different DPL concept?
The FairMQ control plugin pushes configuration options, such as channel configuration, to the device during the `INIT DEVICE` transition (using `SetProperty`). Generally channels are **not** expected to be known at launch time, only at runtime. These are stored in the FMQ options map. We're also looking to push **all** configuration there, but we presently lack a way for user code to query all kinds of configuration from a single user facing API (I've been told that direct access to the FMQ options map isn't pretty from DPL devices). We would like a situation in which the same binary can acquire its configuration the same way regardless of whether its running in AliECS or standalone, and we expected that FMQ options would be the underlying storage for this.
My question is, where does `--configuration ...` end up? Is the payload made accessible in user code with the `Configuration` library? Do the present user code additions allow to query the FMQ options in the same way as `--configuration ...` options, or are these completely separate configuration mechanisms?

**Comment by ktf** [comment]
Yes, it's the FairMQ state machine I am talking about. In particular the `FairMQDevice::Init()` method.

They end up in `ConfigParamRegistry` if the user declares them as ConfigParamSpec in the DataProcessor definition. This can be accessed with:

ctx.options().get<T>("label");

from the Init callback of DPL. I can also hook directly in FairMQ options map (which is how it worked so far), but my understanding is that this needs support  on your side to inject the parameters in some sort of FairMQ plugin, no? Or are you asking WP4 to provide such a plugin?

**Comment by teo** [comment]
@ktf Not at all, the OCC/OCClite plugins can and do already push configuration entries into FairMQOptions, so no additional FairMQ plugin is required from WP4.

I'm paraphrasing a question I commonly get from task developers who run their tasks with AliECS: I have my FairMQ/DPL task and my configuration file. It all runs fine without AliECS. How do I write my configuration access code in such a way that I have full access to the contents of my configuration payload, regardless of whether it comes from a file or from AliECS (INIT DEVICE/FairMQOptions).

This concerns AliECS insofar as I need to know whether the config payload (text document generated by AliECS from a Consul-stored template, can be JSON, INI or anything else) should be parsed and added into the root of the FairMQOptions map key by key (so same as channel configuration), or perhaps dumped into a `boost::ptree` and then added whole under a single key, or just dumped into a single key without processing, in its original format whichever it may be (INI, JSON, etc.) (these are just examples, I don't think all of them are equally a good idea). Regardless of which interface ends up as the single blessed way for DPL users to query their application-specific configuration, AliECS needs to know where to push these payloads to make them available to this interface so the interface can guarantee equal behavior independent of the configuration source.

File: Framework/Core/src/ConfigurationOptionsRetriever.cxx (hunk 1 above)
Comment (awegrzyn) [reply Framework/Core/src/ConfigurationOptionsRetriever.cxx:43]: I don't think so, this is limited by Consul backend as it anyway requires request in order to find out if node exists or not

**Comment by ktf** [comment]
For DPL tasks, they need to declare the options as `ConfigParamSpec` and retrieve them from the init callback (we can make a snapshot available also in the processing callback, if it simplifies things). This guarantees that DPL will use whatever mechanism you specify in the `--configuration` to retrieve the options and defaults to FairMQOptions (now also Configuration, if you prefer). DPL does not support (by design) reading an ini file or any other "out of band" mechanism. Unfortunately I cannot veto people doing it behind DPL back (I would, if I could ;-)).

**Comment by shahor02** [comment]
@ktf is it normal that the options are interfering as ambiguous? This happens with your new "--configuration" and my old "--conf", this workflow was working before:
````
o2-raw-file-reader-workflow --nocheck-missing-stop --delay 0 --conf raw.cfg 
...
[20904:internal-dpl-clock]: [INFO] Spawing new device internal-dpl-clock in process with pid 20904
[20904:internal-dpl-clock]: [ERROR] Unhandled exception reached the top of main, device shutting down. Details follow: 
[20904:internal-dpl-clock]: Throw location unknown (consider using BOOST_THROW_EXCEPTION)
[20904:internal-dpl-clock]: Dynamic exception type: boost::wrapexcept<boost::program_options::ambiguous_option>
[20904:internal-dpl-clock]: std::exception::what: option '--conf' is ambiguous and matches '--config-key', and '--configuration'
[20904:internal-dpl-clock]: 
...
[20905:raw-file-reader]: [INFO] Spawing new device raw-file-reader in process with pid 20905
[20905:raw-file-reader]: [ERROR] Unhandled exception reached the top of main, device shutting down. Details follow: 
[20905:raw-file-reader]: Throw location unknown (consider using BOOST_THROW_EXCEPTION)
[20905:raw-file-reader]: Dynamic exception type: boost::wrapexcept<boost::program_options::ambiguous_option>
[20905:raw-file-reader]: std::exception::what: option '--conf' is ambiguous and matches '--config-key', and '--configuration'
[20905:raw-file-reader]: 
...
````

**Comment by ktf** [comment]
Damn. I think this is due to the fact boost actually treats `--configuration` and `--conf` as the same option (but I bet also `--con`, `--co` etc...). For now I suggest we rename `--configuration` to `--dpl-configuration` or something. Will do so at some point after dinner.

**Comment by shahor02** [comment]
As you prefer, I will also rename my option to more specific. But perhaps we should use disable ``allow_guessing`` (by default?) in the boost options.

**Comment by ktf** [comment]
Probably a good idea. In any case other backends would not support that.

**Pull request merged**